File: client/src/main/java/screen/AchievementsDialog.java
package screen;

import object.AchievementBadge;
import object.RewardStar;

import javax.swing.*;
import javax.swing.border.EmptyBorder;
import java.awt.*;
import java.awt.event.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;

import static screen.AchievementsPanelKt.makeAchievementPanels;
import static util.AchievementUtilKt.getAchievementsEarned;

public class AchievementsDialog extends JFrame 
								implements MouseMotionListener,
										   MouseListener,
										   ActionListener
								
{
	private static final ImageIcon unlockedRewardIcon = new ImageIcon(AchievementsDialog.class.getResource("/rewards/unlockedReward.png"));
	private static final ImageIcon unlockedRewardIconDark = new ImageIcon(AchievementsDialog.class.getResource("/rewards/unlockedRewardDark.png"));
	private static final ImageIcon unlockedRewardIconDarkest = new ImageIcon(AchievementsDialog.class.getResource("/rewards/unlockedRewardDarkest.png"));
	private static final ImageIcon lockedRewardIcon = new ImageIcon(AchievementsDialog.class.getResource("/rewards/lockedReward.png"));

	private int currentPage = 0;
	private int progressShowing = 0;
	
	private boolean redrawing = false;
	
	private ArrayList<AchievementBadge> achievementBadges = new ArrayList<>();
	private ArrayList<AchievementBadge> bonusBadges = new ArrayList<>();
	private ArrayList<RewardStar> rewardStars = new ArrayList<>();
	
	public AchievementsDialog() 
	{
		setTitle("Achievements");
		setSize(860, 504);
		addMouseMotionListener(this);
		setIconImage(new ImageIcon(AchievementsDialog.class.getResource("/icons/achievements.png")).getImage());
		getContentPane().setLayout(null);
		separator.setBounds(0, 342, 878, 2);
		getContentPane().add(separator);
		title.setHorizontalAlignment(SwingConstants.CENTER);
		title.setFont(new Font("Tahoma", Font.BOLD, 20));
		title.setBounds(255, 16, 235, 25);
		getContentPane().add(title);
		panelDescriptions.setBounds(10, 345, 714, 131);
		getContentPane().add(panelDescriptions);
		panelDescriptions.setLayout(null);
		achievementName.setBounds(0, 0, 714, 36);
		achievementName.setFont(new Font("Tahoma", Font.BOLD, 15));
		panelDescriptions.add(achievementName);
		achievementExplanation.setBounds(0, 32, 714, 36);
		panelDescriptions.add(achievementExplanation);
		achievementDescription.setVerticalAlignment(SwingConstants.TOP);
		achievementDescription.setBounds(0, 70, 714, 54);
		achievementDescription.setFont(new Font("Tahoma", Font.ITALIC, 12));
		panelDescriptions.add(achievementDescription);

		reward5.setBounds(820, 287, 17, 16);
		getContentPane().add(reward5);
		reward10.setBounds(715, 272, 17, 16);
		getContentPane().add(reward10);
		reward15.setBounds(820, 257, 17, 16);
		getContentPane().add(reward15);
		reward20.setBounds(715, 242, 17, 16);
		getContentPane().add(reward20);
		reward25.setBounds(820, 227, 17, 16);
		getContentPane().add(reward25);
		reward30.setBounds(715, 212, 17, 16);
		getContentPane().add(reward30);
		reward35.setBounds(820, 197, 17, 16);
		getContentPane().add(reward35);
		reward40.setBounds(715, 182, 17, 16);
		getContentPane().add(reward40);
		reward45.setBounds(820, 167, 17, 16);
		getContentPane().add(reward45);
		reward50.setBounds(715, 152, 17, 16);
		getContentPane().add(reward50);
		testTube.setIcon(new ImageIcon(AchievementsDialog.class.getResource("/tubes/t28.png")));
		testTube.setBounds(730, -28, 92, 344);
		getContentPane().add(testTube);
		btnLeft.setBackground(Color.WHITE);
		btnLeft.setOpaque(false);
		btnLeft.setFont(new Font("Segoe UI Symbol", Font.BOLD, 24));
		btnLeft.setBorder(new EmptyBorder(0, 0, 0, 0));
		btnLeft.setBounds(11, 158, 30, 50);
		getContentPane().add(btnLeft);
		btnLeft.addActionListener(this);
		btnRight.setOpaque(false);
		btnRight.setBackground(Color.WHITE);
		btnRight.setFont(new Font("Segoe UI Symbol", Font.BOLD, 24));
		btnRight.setBorder(new EmptyBorder(0, 0, 0, 0));
		btnRight.setBounds(680, 158, 30, 50);
		getContentPane().add(btnRight);
		btnRight.addActionListener(this);

		btnLeft.setEnabled(false);

		for (AchievementsPanel page : pages) {
			page.setBounds(56, 56, 615, 255);
			getContentPane().add(page);
		}

		populateBadgeListAndAddMotionListeners();
		populateStarListAndAddMouseListeners();
	}
	
	private final JPanel panelDescriptions = new JPanel();
	private final JSeparator separator = new JSeparator();

	private final List<AchievementsPanel> pages = makeAchievementPanels();
	private final JButton btnLeft = new JButton("\u25C0");
	private final JButton btnRight = new JButton("\u25B6");
	
	private final JLabel testTube = new JLabel("");
	private final RewardStar reward5 = new RewardStar(5, "Four Colours", RewardDialog.REWARD_BANNER_FOUR_COLOUR);
	private final RewardStar reward10 = new RewardStar(10, "Negative Jacks", RewardDialog.REWARD_BANNER_NEGATIVE_JACKS);
	private final RewardStar reward15 = new RewardStar(15, "Blind Play", RewardDialog.REWARD_BANNER_BLIND);
	private final RewardStar reward20 = new RewardStar(20, "New Deck Design", RewardDialog.REWARD_BANNER_MINIMALIST);
	private final RewardStar reward25 = new RewardStar(25, "Vectropy", RewardDialog.REWARD_BANNER_VECTROPY);
	private final RewardStar reward30 = new RewardStar(30, "Card Reveal", RewardDialog.REWARD_BANNER_CARD_REVEAL);
	private final RewardStar reward35 = new RewardStar(35, "Extra Suits", RewardDialog.REWARD_BANNER_EXTRA_SUITS);
	private final RewardStar reward40 = new RewardStar(40, "Illegal!", RewardDialog.REWARD_BANNER_ILLEGAL);
	private final RewardStar reward45 = new RewardStar(45, "New Joker Design", RewardDialog.REWARD_BANNER_DEVELOPERS);
	private final RewardStar reward50 = new RewardStar(50, "Cheats", RewardDialog.REWARD_BANNER_CHEATS);
	
	private final JLabel title = new JLabel("");
	private final JLabel achievementName = new JLabel("Achievement Name");
	private final JLabel achievementExplanation = new JLabel("How you earned it");
	private final JLabel achievementDescription = new JLabel("Humorous description");
	
	public void init()
	{
		achievementName.setText("");
		achievementDescription.setText("");
		achievementExplanation.setText("");

		updatePagination();
		redrawStars();
		animateTestTube();
	}
	
	private void populateBadgeListAndAddMotionListeners()
	{
		pages.forEach(page -> {
			for (AchievementBadge badge : page.getBadges())
			{
				badge.addMouseMotionListener(this);
			}
		});

	}
	
	private void populateStarListAndAddMouseListeners()
	{
		Component[] components = getContentPane().getComponents();
		for (int i=0; i<components.length; i++)
		{
			Component c = components[i];
			if (c instanceof RewardStar)
			{
				RewardStar star = (RewardStar)c;
				star.addMouseListener(this);
				rewardStars.add(star);
			}
		}
	}
	
	private void animateTestTube()
	{
		redrawing = true;
		
		Timer timer = new Timer("Timer-TestTube");
		progressShowing = 0;
		testTube.setIcon(getTubeIconForIndex(0));
		
		for (int i = 1; i<= getAchievementsEarned(); i++)
		{
			timer.schedule(new RedrawTestTube(), 30 * i);
		}
	}
	
	private void updateTestTube()
	{
		testTube.setIcon(getTubeIconForIndex(getAchievementsEarned()));
	}
	
	private ImageIcon getTubeIconForIndex(int i)
	{
		String name = "t" + i + ".png";
		return new ImageIcon(AchievementsDialog.class.getResource("/tubes/" + name));
	}
	
	private void redrawStars()
	{
		for (int i=0; i<rewardStars.size(); i++)
		{
			RewardStar star = rewardStars.get(i);
			
			if (star.isUnlocked(progressShowing))
			{
				star.setIcon(unlockedRewardIcon);
				String hoverDesc = star.getHoverDesc();
				star.setToolTipText(hoverDesc);
			}
			else
			{
				star.setIcon(lockedRewardIcon);
				star.setToolTipText("Locked");
			}
		}
	}
	
	public void refresh(boolean restartTube)
	{
		if (restartTube)
		{
			animateTestTube();
		}
		else
		{
			updateTestTube();
		}
	}

	@Override
	public void mouseDragged(MouseEvent arg0) {}

	@Override
	public void mouseMoved(MouseEvent arg0) 
	{
		Component c = arg0.getComponent();
		
		if (c instanceof AchievementBadge)
		{
			AchievementBadge badge = (AchievementBadge)arg0.getComponent();
			achievementName.setText(c.getName());
			
			if (badge.isEarned())
			{
				achievementExplanation.setText(badge.getExplanation());
				achievementDescription.setText(badge.getDescription());

			}
			else
			{
				achievementExplanation.setText("???");
				achievementDescription.setText("");
			}
		}
		else
		{
			achievementName.setText("");
			achievementExplanation.setText("");
			achievementDescription.setText("");
		}
	}
	
	private class RedrawTestTube extends TimerTask
	{
		@Override
		public void run() 
		{
			if (isVisible())
			{
				if (progressShowing < getAchievementsEarned())
				{
					progressShowing++;
				}
				
				testTube.setIcon(getTubeIconForIndex(progressShowing));
				redrawStars();
				
				if (progressShowing == getAchievementsEarned())
				{
					redrawing = false;
				}
			}
		}
	}

	@Override
	public void mouseClicked(MouseEvent e) 
	{
		RewardStar c = (RewardStar)e.getComponent();
		if (c.isUnlocked(getAchievementsEarned())
		  && !redrawing)
		{
			String imageName = c.getImageName();
			RewardDialog.showDialog(imageName);
		}
	}

	@Override
	public void mouseEntered(MouseEvent e) 
	{
		RewardStar source = (RewardStar)e.getComponent();
		if (source.isUnlocked(progressShowing) && !redrawing)
		{
			source.setIcon(unlockedRewardIconDark);
		}
	}

	@Override
	public void mouseExited(MouseEvent e) 
	{
		RewardStar source = (RewardStar)e.getComponent();
		if (source.isUnlocked(progressShowing) && !redrawing)
		{
			source.setIcon(unlockedRewardIcon);
		}
	}
	
	@Override
	public void mousePressed(MouseEvent e) 
	{
		RewardStar source = (RewardStar)e.getComponent();
		if (source.isUnlocked(getAchievementsEarned()) && !redrawing)
		{
			source.setIcon(unlockedRewardIconDarkest);
		}
	}
	
	@Override
	public void mouseReleased(MouseEvent e) 
	{
		RewardStar source = (RewardStar)e.getComponent();
		if (source.isUnlocked(getAchievementsEarned()) && !redrawing)
		{
			source.setIcon(unlockedRewardIcon);
		}
	}

	@Override
	public void actionPerformed(ActionEvent arg0)
	{
		if (arg0.getSource() == btnLeft) {
			currentPage--;
		} else {
			currentPage++;
		}

		updatePagination();
	}

	private void updatePagination() {
		btnLeft.setEnabled(currentPage > 0);
		btnRight.setEnabled(currentPage < pages.size());

		for (int i=0; i<pages.size(); i++) {
			var page = pages.get(i);
			page.setVisible(i == currentPage);

			if (i == currentPage) {
				title.setText(page.getTitle());
			}
		}
	}
}

File: client/src/main/kotlin/screen/AchievementsDialog.kt
package screen

import bean.AchievementBadge
import java.awt.Color
import java.awt.Font
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.awt.event.MouseEvent
import java.awt.event.MouseListener
import java.awt.event.MouseMotionListener
import java.util.*
import javax.swing.ImageIcon
import javax.swing.JButton
import javax.swing.JFrame
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JSeparator
import javax.swing.SwingConstants
import javax.swing.border.EmptyBorder
import `object`.RewardStar
import util.Images
import util.getAchievementsEarned
import utils.getAllChildComponentsForType

class AchievementsDialog : JFrame(), MouseMotionListener, MouseListener, ActionListener {
    private var currentPage = 0
    private var progressShowing = 0

    private var redrawing = false

    private val panelDescriptions = JPanel()
    private val separator = JSeparator()

    private val pages = makeAchievementPanels()
    private val btnLeft = JButton("\u25C0")
    private val btnRight = JButton("\u25B6")

    private val testTube = JLabel("")
    private val reward5: RewardStar =
        RewardStar(5, "Four Colours", RewardDialog.REWARD_BANNER_FOUR_COLOUR)
    private val reward10: RewardStar =
        RewardStar(10, "Negative Jacks", RewardDialog.REWARD_BANNER_NEGATIVE_JACKS)
    private val reward15: RewardStar =
        RewardStar(15, "Blind Play", RewardDialog.REWARD_BANNER_BLIND)
    private val reward20: RewardStar =
        RewardStar(20, "New Deck Design", RewardDialog.REWARD_BANNER_MINIMALIST)
    private val reward25: RewardStar =
        RewardStar(25, "Vectropy", RewardDialog.REWARD_BANNER_VECTROPY)
    private val reward30: RewardStar =
        RewardStar(30, "Card Reveal", RewardDialog.REWARD_BANNER_CARD_REVEAL)
    private val reward35: RewardStar =
        RewardStar(35, "Extra Suits", RewardDialog.REWARD_BANNER_EXTRA_SUITS)
    private val reward40: RewardStar =
        RewardStar(40, "Illegal!", RewardDialog.REWARD_BANNER_ILLEGAL)
    private val reward45: RewardStar =
        RewardStar(45, "New Joker Design", RewardDialog.REWARD_BANNER_DEVELOPERS)
    private val reward50: RewardStar = RewardStar(50, "Cheats", RewardDialog.REWARD_BANNER_CHEATS)

    private val title = JLabel("")
    private val achievementName = JLabel("Achievement Name")
    private val achievementExplanation = JLabel("How you earned it")
    private val achievementDescription = JLabel("Humorous description")

    init {
        setTitle("Achievements")
        setSize(860, 504)
        addMouseMotionListener(this)
        iconImage =
            ImageIcon(AchievementsDialog::class.java.getResource("/icons/achievements.png")).image
        contentPane.layout = null
        separator.setBounds(0, 342, 878, 2)
        contentPane.add(separator)
        title.horizontalAlignment = SwingConstants.CENTER
        title.font = Font("Tahoma", Font.BOLD, 20)
        title.setBounds(255, 16, 235, 25)
        contentPane.add(title)
        panelDescriptions.setBounds(10, 345, 714, 131)
        contentPane.add(panelDescriptions)
        panelDescriptions.layout = null
        achievementName.setBounds(0, 0, 714, 36)
        achievementName.font = Font("Tahoma", Font.BOLD, 15)
        panelDescriptions.add(achievementName)
        achievementExplanation.setBounds(0, 32, 714, 36)
        panelDescriptions.add(achievementExplanation)
        achievementDescription.verticalAlignment = SwingConstants.TOP
        achievementDescription.setBounds(0, 70, 714, 54)
        achievementDescription.font = Font("Tahoma", Font.ITALIC, 12)
        panelDescriptions.add(achievementDescription)

        reward5.setBounds(820, 287, 17, 16)
        contentPane.add(reward5)
        reward10.setBounds(715, 272, 17, 16)
        contentPane.add(reward10)
        reward15.setBounds(820, 257, 17, 16)
        contentPane.add(reward15)
        reward20.setBounds(715, 242, 17, 16)
        contentPane.add(reward20)
        reward25.setBounds(820, 227, 17, 16)
        contentPane.add(reward25)
        reward30.setBounds(715, 212, 17, 16)
        contentPane.add(reward30)
        reward35.setBounds(820, 197, 17, 16)
        contentPane.add(reward35)
        reward40.setBounds(715, 182, 17, 16)
        contentPane.add(reward40)
        reward45.setBounds(820, 167, 17, 16)
        contentPane.add(reward45)
        reward50.setBounds(715, 152, 17, 16)
        contentPane.add(reward50)
        testTube.icon = ImageIcon(AchievementsDialog::class.java.getResource("/tubes/t28.png"))
        testTube.setBounds(730, -28, 92, 344)
        contentPane.add(testTube)
        btnLeft.background = Color.WHITE
        btnLeft.isOpaque = false
        btnLeft.font = Font("Segoe UI Symbol", Font.BOLD, 24)
        btnLeft.border = EmptyBorder(0, 0, 0, 0)
        btnLeft.setBounds(25, 158, 30, 50)
        contentPane.add(btnLeft)
        btnLeft.addActionListener(this)
        btnRight.isOpaque = false
        btnRight.background = Color.WHITE
        btnRight.font = Font("Segoe UI Symbol", Font.BOLD, 24)
        btnRight.border = EmptyBorder(0, 0, 0, 0)
        btnRight.setBounds(645, 158, 30, 50)
        contentPane.add(btnRight)
        btnRight.addActionListener(this)

        btnLeft.isEnabled = false

        for (page in pages) {
            page.setBounds(56, 56, 590, 255)
            contentPane.add(page)
        }

        populateBadgeListAndAddMotionListeners()
        populateStarListAndAddMouseListeners()
    }

    fun init() {
        achievementName.text = ""
        achievementDescription.text = ""
        achievementExplanation.text = ""

        updatePagination()
        refresh(true)
    }

    private fun populateBadgeListAndAddMotionListeners() {
        getAllChildComponentsForType<AchievementBadge>().forEach { it.addMouseMotionListener(this) }
    }

    private fun populateStarListAndAddMouseListeners() {
        getAllChildComponentsForType<RewardStar>().forEach { it.addMouseListener(this) }
    }

    private fun animateTestTube() {
        redrawing = true

        val timer = Timer("Timer-TestTube")
        progressShowing = 0
        testTube.icon = getTubeIconForIndex(0)

        for (i in 1..getAchievementsEarned()) {
            timer.schedule(RedrawTestTube(), (30 * i).toLong())
        }
    }

    private fun updateTestTube() {
        progressShowing = getAchievementsEarned()
        testTube.icon = getTubeIconForIndex(getAchievementsEarned())
        redrawStars()
    }

    private fun getTubeIconForIndex(i: Int): ImageIcon {
        val name = "t$i.png"
        return ImageIcon(AchievementsDialog::class.java.getResource("/tubes/$name"))
    }

    private fun redrawStars() {
        getAllChildComponentsForType<RewardStar>().forEach { star ->
            if (star.isUnlocked(getAchievementsEarned())) {
                star.setIcon(Images.REWARD_UNLOCKED)
                star.setToolTipText(star.hoverDesc)
            } else {
                star.setIcon(Images.REWARD_LOCKED)
                star.setToolTipText("Locked")
            }

            star.repaint()
        }
    }

    fun refresh(restartTube: Boolean) {
        if (restartTube) {
            animateTestTube()
        } else {
            updateTestTube()
        }

        getAllChildComponentsForType<AchievementBadge>().forEach { it.toggle() }

        updateTitle()
        repaint()
    }

    override fun mouseDragged(arg0: MouseEvent) {}

    override fun mouseMoved(arg0: MouseEvent) {
        val c = arg0.component

        if (c is AchievementBadge) {
            val badge: AchievementBadge = arg0.component as AchievementBadge
            achievementName.text = c.name

            if (badge.isEarned) {
                achievementExplanation.text = badge.explanation
                achievementDescription.text = badge.description
            } else {
                achievementExplanation.text = "???"
                achievementDescription.text = ""
            }
        } else {
            achievementName.text = ""
            achievementExplanation.text = ""
            achievementDescription.text = ""
        }
    }

    private inner class RedrawTestTube : TimerTask() {
        override fun run() {
            if (isVisible) {
                if (progressShowing < getAchievementsEarned()) {
                    progressShowing++
                }

                testTube.icon = getTubeIconForIndex(progressShowing)
                redrawStars()

                if (progressShowing == getAchievementsEarned()) {
                    redrawing = false
                }
            }
        }
    }

    override fun mouseClicked(e: MouseEvent) {
        val c: RewardStar = e.component as RewardStar
        if (c.isUnlocked(getAchievementsEarned()) && !redrawing) {
            RewardDialog.showDialog(c.imageName)
        }
    }

    override fun mouseEntered(e: MouseEvent) {
        val source: RewardStar = e.component as RewardStar
        if (source.isUnlocked(progressShowing) && !redrawing) {
            source.setIcon(Images.REWARD_UNLOCKED_HOVERED)
        }
    }

    override fun mouseExited(e: MouseEvent) {
        val source: RewardStar = e.component as RewardStar
        if (source.isUnlocked(progressShowing) && !redrawing) {
            source.setIcon(Images.REWARD_UNLOCKED)
        }
    }

    override fun mousePressed(e: MouseEvent) {
        val source: RewardStar = e.component as RewardStar
        if (source.isUnlocked(getAchievementsEarned()) && !redrawing) {
            source.setIcon(Images.REWARD_UNLOCKED_CLICKED)
        }
    }

    override fun mouseReleased(e: MouseEvent) {
        val source: RewardStar = e.component as RewardStar
        if (source.isUnlocked(getAchievementsEarned()) && !redrawing) {
            source.setIcon(Images.REWARD_UNLOCKED)
        }
    }

    override fun actionPerformed(arg0: ActionEvent) {
        if (arg0.source === btnLeft) {
            currentPage--
        } else {
            currentPage++
        }

        updatePagination()
    }

    private fun updatePagination() {
        btnLeft.isEnabled = currentPage > 0
        btnRight.isEnabled = currentPage < pages.size - 1

        pages.forEachIndexed { i, page -> page.isVisible = i == currentPage }
        updateTitle()
    }

    private fun updateTitle() {
        title.text = pages[currentPage].getTitle()
    }
}
